Integracion insightMetrics

rafgod opened the pull request:
## Cambios realizados para el feature:
 * Se agregaron algunos parametros mas para armar el request de metrics
 * Se hicieron casos de pruebas de llamado a insightLight y metrics

## Cambios generales
 * Integracion Insight
 
## Issues cerrados
 * Closes #XX
 

## Checklist validación PR:

- [* ] Título y descripción clara del PR
- [* ] Tests de mi funcionalidad 
- [ ] Documentación de mi funcionalidad
- [ *] Tests ejecutados y pasados
- [ ] Branch Coverage >= 80%

rafgod-ML committed:
Se agregaron mas parametros al request de insight

rafgod-ML committed:
Se agregaron Unit test de insight

rafgod-ML committed:
correccion de observaciones

rafgod-ML committed:
Cambio de identacion

Cambio de identacion

rafgod-ML committed:
Merge branch 'feature/SRE-1981'

merge master

rafgod-ML committed:
cambio observaciones PR y merge master

rafgod-ML committed:
cambio observaciones PR y merge master/ merge

rafgod-ML committed:
borre SYSO de test

delias-silva committed:
Release connection in insight api calls.

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,150 @@
+import org.apache.http.util.EntityUtils;
+
+
+public class InsightDataManager {
```
esta clase debería ser un singleton y tener la responsabilidad de llamar al back-end para:
- consultar traffic-light
- enviar métricas

además exponer métodos de instancia (no estáticos) para ser consultados por stats() que digan:
- si esta o no habilitada para mandar métricas
- si un endpoint esta o no en la white-list
- enviar métricas

rafgod-ML replied:
Listo

rkuzner commented on src/main/java/com/mercadopago/insight/InsightConnectionManager.java:
```diff
@@ -0,0 +1,84 @@
+import org.apache.http.message.BasicHttpResponse;
+import org.apache.http.message.BasicStatusLine;
+
+public class InsightConnectionManager {
```
tal como hablamos, debe ir dentro de `InsightsDataManager()`

rkuzner commented on src/main/java/com/mercadopago/insight/Stats.java:
```diff
@@ -0,0 +1,76 @@
+
+public class Stats extends Thread {
+
+    static final String INSIGHT_DEFAULT_BASE_URL = "https://events.mercadopago.com/";
```
ya que estamos, las constantes se van dentro de `InsightsDataManager()`

rkuzner commented on src/main/java/com/mercadopago/insight/Stats.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    public static boolean isEndpointInWhiteList(TrafficLightResponse trafficLight, String endpoint) {
```
mover este método a público de instancia de `InsightDataManager()`

rkuzner commented on src/main/java/com/mercadopago/insight/TrafficLightManager.java:
```diff
@@ -0,0 +1,77 @@
+import org.apache.http.protocol.HTTP;
+import org.apache.http.util.EntityUtils;
+
+public class TrafficLightManager {
```
tal como hablamos, debe ir dentro de `InsightsDataManager()`

rkuzner commented on src/main/java/com/mercadopago/insight/TrafficLightManager.java:
```diff
@@ -0,0 +1,77 @@
+    private static void getDefaultResponse(){
+        trafficLightResponse = new TrafficLightResponse();
+        trafficLightResponse.setSendDataEnabled(false);
+    }
```
incluir el TTL para que ante una falla solo consulte nuevamente al back-end luego de 10 minutos (600 segundos)

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,150 @@
+
+            TcpInfo tcpInfo = new TcpInfo.Builder().withSourceAddress(getHostAddress()).build();
+
+            ConnectionInfo connectionInfo = new ConnectionInfo.Builder().withUserAgent(request.getLastHeader(HTTP.USER_AGENT)!=null?request.getLastHeader(HTTP.USER_AGENT).getValue():"")
```
solo usaría `withUserAgent()` cuando tengo valor y no guardaría string vacío.

rafgod-ML committed:
Correccion de observaciones 

Correccion de observaciones añadidas por Raul al PR

rafgod-ML committed:
cambio de ambiente a prod

rafgod-ML committed:
Agregue un defaultResponse

Agregue un defauult response para trafficLight en caso de que ocurra una exception en el llamado a la API

rafgod-ML committed:
Correcion del TTL para no ir siempre a preguntar en caso de error

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+    private static long sendDataDeadlineMillis = Long.MIN_VALUE;
+    private static InsightDataManager insightDataManager = null;
+
+    public static  InsightDataManager getInsightDataManager(){
```
nombre sugerido `getInstance()` (por singleton pattern)

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+     * Method to ask the back-end if the client should send Insights Data (or not).
+     * @return HttpResponse 
+     */
+    public HttpResponse callTrafficLight() {
```
no debería ser `public`

rafgod-ML replied:
listo

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+            HttpEntity entityRes = lightResponse.getEntity();
+            if (entityRes != null) {
+                trafficLight = new Gson().fromJson(EntityUtils.toString(entityRes), TrafficLightResponse.class);
+                sendDataDeadlineMillis = System.currentTimeMillis() + (Math.abs(trafficLight.getSendTTL()) * 1000);
```
esto se puede hacer fuera del try-catch y de paso tampoco lo hace getDefaultResponse()

rafgod-ML replied:
listo

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+        return insightResponse;
+    }
+
+    public Boolean isInsightMetricsEnable(String url){
```
le agregaría el comentario sobre la lógica de decisión sobre si la url está o no en la white-list.
(traer el comment del modulo de android)

rafgod-ML replied:
listo

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+    static final String HEADER_X_INSIGHTS_EVENT_NAME = "X-Insights-Event-Name";
+    static final String HEADER_X_INSIGHTS_METRIC_LAB_SCOPE = "X-Insights-Metric-Lab-Scope";
+    static final String HEADER_X_INSIGHTS_DATA_ID = "X-Insights-Data-Id";
+    static final String HEADER_X_INSIGHTS_DATA = "X-Insights-Data";
```
como tenemos un httpClient exclusivo para hablar con el backend, este header no lo necesitamos

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+           
+            ClientInfo clientInfo = new ClientInfo.Builder().withName(MercadoPago.SDK.getClientName()).withVersion(MercadoPago.SDK.getVersion()).build();
+
+            BusinessFlowInfo.Builder businessFlowInfoBuilder = new BusinessFlowInfo.Builder()
```
para `business-flow-info`, ver como está resuelto en https://github.com/mercadolibre/fury_mp-config-provider-android/blob/develop/networking-configurer/src/main/java/com/mercadolibre/networking_configurer/restclient/configurator/decorator/client/insights/InsightsDataManager.java#L595

rafgod-ML replied:
listo

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+                businessFlowInfoBuilder.withName(request.getLastHeader(HEADER_X_INSIGHTS_BUSINESS_FLOW).getValue());
+            }
+
+            EventInfo eventInfo = new EventInfo.Builder().withName(request.getLastHeader(HEADER_X_INSIGHTS_EVENT_NAME)!=null?request.getLastHeader(HEADER_X_INSIGHTS_EVENT_NAME).getValue():"Unknown").build();
```
solo mandaría `event-info` si `event-name` tiene valor

rafgod-ML replied:
listo

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -0,0 +1,354 @@
+		return response;
+    }
+    
+    public TrafficLightResponse getTrafficLightResponse(){
```
no debería ser `public`

rafgod-ML committed:
Cambio por observaciones de PR

rafgod-ML committed:
uso de constante para eventName

rafgod-ML committed:
formato de codigo

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -264,40 +297,75 @@ public boolean isEndpointInWhiteList(TrafficLightResponse trafficLight, String e
+            String osName = System.getProperty("os.name")!=null?System.getProperty("os.name"):"";
+            String osVersion = System.getProperty("os.version")!=null?System.getProperty("os.version"):"";
+            String deviceRam = getMemorySize();
+            String modelName = System.getProperty("os.arch")!=null?System.getProperty("os.arch"):"";
```
este valor va sobre el cpu-type junto a cantidad de CPUs

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -264,40 +297,75 @@ public boolean isEndpointInWhiteList(TrafficLightResponse trafficLight, String e
 
-    private String getHostAddress(){
-        InetAddress localhost = null;
+    private static void getDefaultResponse() {
```
no debe ser `static`

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -264,40 +297,75 @@ public boolean isEndpointInWhiteList(TrafficLightResponse trafficLight, String e
-        String osName = System.getProperty("os.name")!=null?System.getProperty("os.name"):"Unknown";
-        String osVersion = System.getProperty("os.version")!=null?System.getProperty("os.version"):"Unknown";
-        return osName + " "+ osVersion;
+    private static String getMemorySize() {
```
no es necesario que sea `static`

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -264,40 +297,75 @@ public boolean isEndpointInWhiteList(TrafficLightResponse trafficLight, String e
+        DeviceInfo.Builder deviceInfo = new DeviceInfo.Builder();
         try {
-            localhost = InetAddress.getLocalHost();
+            int availableCPU = Runtime.getRuntime().availableProcessors();
```
todos los valores del dispositivo podrías calcularlos una sola vez y guardarlos en variables privadas para luego usarlos todas la veces que necesites

rafgod-ML committed:
ultimas observaciones

rkuzner commented on src/main/java/com/mercadopago/insight/InsightDataManager.java:
```diff
@@ -301,48 +314,41 @@ public boolean isEndpointInWhiteList(TrafficLightResponse trafficLight, String r
         try {
             long memSize = ((com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean()).getTotalPhysicalMemorySize();
-            memorySize = "RAM Size=" + memSize + " Bytes";
+            memorySize = "RAM Size " + memSize + " Bytes";
```
por favor, mandá solo el valor en bytes (sin el texto de "ram size" ni "bytes")

rafgod-ML committed:
se elimino la leyenda ram size
